### PR TITLE
Adding details to Conformance GEP

### PR DESCRIPTION
**Pull request by robscott**
**What type of PR is this?**
/kind documentation
/kind gep

**What this PR does / why we need it**:
This is an addition to GEP #917 that provides a general structure for how we can define conformance tests. Although this could have potentially been a separate GEP, it felt sufficiently small to just bundle with the existing conformance GEP. Happy to split this out if it's easier though.

This is largely a follow up to the [earlier doc](https://docs.google.com/document/d/1QL-MpIVzqxe32Y2BZ_dYOB8zNsF9c4pnKEIB9ZLt118/edit), but it also adds a section to discuss how this could work with prepopulated Gateways. Open to other approaches here, this just seemed like it may be the simplest starting point.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Comment by robscott** [comment]
This is adding to Nick's GEP, so adding him to reviewers list as well:

/cc @youngnick

**Comment by youngnick** [comment]
Yes, I agree with the approach here, I feel like it will put some pressure on those of us with currently-non-dynamic Gateway provisioning to get it built (😄 ), but the support for preloaded Gateways will cover us in the meantime.

/lgtm

File: site-src/geps/gep-917.md [hunk 1]
```diff
@@ -126,23 +124,46 @@ Fields marked "Extended" support will eventually have conformance tests that
...
 to tell the testing framework what extended fields they support.
 
-#### Testing framework
+## Testing framework
+
+Conformance tests will initially be structured as a CLI that runs on existing
+Kubernetes clusters. This may be expanded to include a testing library that can
+be imported by implementations. This tool will output test results in a
+structured format that can be used to generate reports. Where possible it will
+share code with the existing ingress-controller-conformance test suite.
```
Comment (shaneutt) [inline site-src/geps/gep-917.md:133]: For your considerations: while I think a CLI is good in any case, I think structuring this as a Kubernetes `Job` as the sort of defacto way to run this may have some merit.

File: site-src/geps/gep-917.md [hunk 2]
```diff
@@ -126,23 +124,46 @@ Fields marked "Extended" support will eventually have conformance tests that
...
+
+### Test Definitions
+
+Each set of tests will be built around a set of YAML manifests representing
+Gateway API resources. The tests will focus on how they expect a given
+controller to process those resources, configured via GatewayClass name. For
+example, tests may check to ensure that a controller correctly populates status,
+or that a request is appropriately routed to the desired backend with any
+relevant modifications.
```
Comment (shaneutt) [inline site-src/geps/gep-917.md:142]: This vaguely reminds me of https://github.com/kudobuilder/kuttl, which I worked on with some friends at a previous job. While I don't necessarily suggest we use `kuttl` itself it might be worth reviewing it for ideas and influences if nothing else.

File: site-src/geps/gep-917.md [hunk 3]
```diff
@@ -126,23 +124,46 @@ Fields marked "Extended" support will eventually have conformance tests that
...
-of tests is a requirment for the API to graduate to `v1beta1` API stability level.
-(As per our upstream KEP). It's acceptable to begin with a small set and expand
-outward while the project is in beta, but having a full set of conformance tests
-that cover most of the API scope should be a requirement for declaring the API
-stable.
+* Docs URL
+* Function to determine feature enablement
 
-#### Certification process
+### Prepopulated Gateways
```
Comment (shaneutt) [inline site-src/geps/gep-917.md:155]: FWIW I'm personally in favor of calling these "unmanaged" gateways as I feel that language is more clear at a glance.

File: site-src/geps/gep-917.md [hunk 4]
```diff
@@ -126,23 +124,46 @@ Fields marked "Extended" support will eventually have conformance tests that
...
+
+Some implementations of Gateway API are not able to provision arbitrary Gateways
+and instead are deployed with a 1:1 relationship to a Gateway. To ensure that
```
Comment (shaneutt) [inline site-src/geps/gep-917.md:158]: While it's basically a nitpick, I think the language here about ability might not be necessary because there's implementations like Istio which have the ability but also support "unmanaged" gateways.

My thoughts:
```suggestion
Some implementations of Gateway API support unmanaged `Gateway` resources 
that refer to an underlying server that was created by other means (e.g. `Helm`)
rather than provisioning and managing the server according to the `Gateway`
resource. To ensure that
```

File: site-src/geps/gep-917.md [hunk 5]
```diff
@@ -1,21 +1,21 @@
...
+# GEP-917: Gateway API Conformance Testing
 
 * Issue: [#917](https://github.com/kubernetes-sigs/gateway-api/issues/917)
 * Status: Provisional
 
 ## TLDR
 
-This GEP outlines the reasons for and principles by which the Gateway API will
-design its conformance and testing regime.
+This GEP outlines the principles and overarching structure that Gateway API will
```
Comment (hbagdi) [inline site-src/geps/gep-917.md:8]:
```suggestion
This GEP outlines the principles and overarching structure that Gateway API
```

**Comment by hbagdi** [comment]
This is reasonable and we should move quickly here. We will learn so much more once we have 2 or 3 conformance tests and a few implementations running those in their respective CIs.
/lgtm

File: site-src/geps/gep-917.md (hunk 1 above)
Comment (robscott) [reply site-src/geps/gep-917.md:133]: Agree that that's a natural extension point, but I think even a k8s job would be based around a CLI.

File: site-src/geps/gep-917.md (hunk 2 above)
Comment (robscott) [reply site-src/geps/gep-917.md:142]: Thanks! I took a look at kuttl and it has some pretty great patterns. I think our conformance tests will need to be somewhat more complex, specifically making potentially advanced requests and ensuring their responses match what we'd expect. This does seem like a great reference point for any of our tests that will start with a set of Kubernetes resources and expect status to be populated in a specific way.

File: site-src/geps/gep-917.md (hunk 3 above)
Comment (robscott) [reply site-src/geps/gep-917.md:155]: Although "unmanaged" makes sense for the most common use case here, I think "prepopulated" is a bit more broad here. For example, some implementations may have very slow provisioning, or some manual steps during initial provisioning followed by automated updates afterwards. What this is really referring to is the ability for implementations to prepopulate Gateways, for any reason, when running conformance tests.

File: site-src/geps/gep-917.md (hunk 4 above)
Comment (robscott) [reply site-src/geps/gep-917.md]: Good clarification - updated.

File: site-src/geps/gep-917.md (hunk 5 above)
Comment (robscott) [reply site-src/geps/gep-917.md]: Updated, thanks!

**Comment by robscott** [comment]
I think there's enough consensus on this one. Removing the hold.

/hold cancel
